src/goto-programs/remove_instanceof.cpp

```diff
@@ -42,160 +42,129 @@ class remove_instanceoft
+  // not contain derefs.
+  // We actually insert the assignment instruction after the existing one.
+  // This will briefly be ill-formed (use before def of instanceof_tmp) but the
+  // two will subsequently switch places. This makes sure that the inserted
+  // assignement doesn't end up before any labels pointing at this instruction.
+  symbol_typet jlo("java::java.lang.Object");
+  exprt object_clsid=get_class_identifier_field(check_ptr, jlo, ns);
+
+  symbolt &newsym=
+    get_fresh_aux_symbol(
+      object_clsid.type(),
+      "instanceof_tmp",
+      "instanceof_tmp",
+      source_locationt(),
+      ID_java,
+      symbol_table);
+
+  auto newinst=goto_program.insert_after(this_inst);
+  newinst->make_assignment();
+  newinst->code=code_assignt(newsym.symbol_expr(), object_clsid);
+  newinst->source_location=this_inst->source_location;
+  newinst->function=this_inst->function;
+
+  // Replace the instanceof construct with a big-or.
+  exprt::operandst or_ops;
+  for(const auto &clsname : children)
   {
-    Forall_operands(it, expr)
-      lower_instanceof(*it, goto_program, this_inst, inst_switch);
+    constant_exprt clsexpr(clsname, string_typet());
+    equal_exprt test(newsym.symbol_expr(), clsexpr);
+    or_ops.push_back(test);
   }
+  expr=disjunction(or_ops);
+
+  return 1;
 }
 
-/// See function above
-/// \par parameters: GOTO program instruction `target` whose instanceof
-///   expressions,
-/// if any, should be replaced with explicit tests, and the
-/// `goto_program` it is part of.
-/// \return Side-effect on `target` as above.
-void remove_instanceoft::lower_instanceof(
+/// Replaces expressions like e instanceof A with e.\@class_identifier == "A"
+/// or a big-or of similar expressions if we know of subtypes that also satisfy
+/// the given test. Does this for the code or guard at a specific instruction.
+/// \param goto_program: program to process
+/// \param target: instruction to check for instanceof expressions
+/// \return true if an instanceof has been replaced
+bool remove_instanceoft::lower_instanceof(
   goto_programt &goto_program,
-  goto_programt::targett target,
-  instanceof_instt &inst_switch)
+  goto_programt::targett target)
 {
-  bool code_iof=contains_instanceof(target->code);
-  bool guard_iof=contains_instanceof(target->guard);
-  // The instruction-switching step below will malfunction if a check
-  // has been added for the code *and* for the guard. This should
-  // be impossible, because guarded instructions currently have simple
-  // code (e.g. a guarded-goto), but this assertion checks that this
-  // assumption is correct and remains true on future evolution of the
-  // allowable goto instruction types.
-  assert(!(code_iof && guard_iof));
-  if(code_iof)
-    lower_instanceof(target->code, goto_program, target, inst_switch);
-  if(guard_iof)
-    lower_instanceof(target->guard, goto_program, target, inst_switch);
+  int replacements=
```

Comment:
use an unsigned type here

Reply:
"Number of" is usually a std::size_t, not unsigned.